### PR TITLE
PLANNER-2562: Add employee scheduling quickstart

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
### JIRA

https://issues.redhat.com/browse/PLANNER-2562

### Referenced pull requests

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add the label `run_fdb`
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] tests</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
* for a <b>specific Native LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] native</b>
</details>

### CODE REVIEW

**Comment by Christopher-Chianelli** [comment]
TODO: Add Continuous planning / Non-disruptive replanning

File: use-cases/employee-scheduling/README.adoc [hunk 1]
```diff
@@ -0,0 +1,99 @@
+= Maintenance Scheduling (Java, Quarkus, Maven)
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/README.adoc]:
```suggestion
= Employee Scheduling (Java, Quarkus, Maven)
```

File: use-cases/employee-scheduling/README.adoc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+Schedule maintenance jobs to crews over time to reduce both premature and overdue maintenance.
+
+image::../../build/quickstarts-showcase/src/main/resources/META-INF/resources/screenshot/quarkus-maintenance-scheduling-screenshot.png[]
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/README.adoc]: these references would need to change too

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/bootstrap/DemoDataGenerator.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    DemoData demoData;
+
+    public enum DemoData {
+        NONE,
+        SMALL,
+        LARGE
+    }
+
+    final String[] REQUIRED_SKILLS = { "Doctor", "Nurse"};
+    final String[] OPTIONAL_SKILLS = { "Anaesthetics", "Cardiology"};
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/bootstrap/DemoDataGenerator.java]: for ALL_CAPS notation, they need to be static too in the typical code convention IIRC. Either way, it wouldn't hurt to make them static too:

static final String[] ...

**Comment by Christopher-Chianelli** [comment]
I think we should add a 8 hours between shifts constraint; I noticed sometimes an afternoon shift is scheduled immediately after a night shift.

File: build/quickstarts-showcase/src/main/resources/META-INF/resources/index.html [hunk 4]
```diff
@@ -189,6 +189,30 @@ <h5>
...
+                            <!-- TODO: Blog Post -->
+                            <a href="https://www.optaplanner.org/blog/2021/10/14/OrderPickingQuickstart.html" target="_blank" title="About this quickstart" class="float-right">
+                                <span class="fas fa-info-circle"></span>
+                            </a>
+                        </h5>
+                    </div>
+                    <img src="screenshot/quarkus-employee-scheduling-screenshot.png" class="card-img-top mt-3" alt="Screenshot"/>
+                    <div class="card-body">
+                        <p class="card-text">
+                            Schedule shifts to employees, accounting for employee availabilities.
```
Comment (ge0ffrey) [inline build/quickstarts-showcase/src/main/resources/META-INF/resources/index.html]: Let's mention skills too. 
```suggestion
                            Schedule shifts to employees, accounting for employee availability and shift skill requirements.
```
Let's use the exact same sentence for the other mentions (README etc).

File: build/quickstarts-showcase/src/main/resources/META-INF/resources/index.html [hunk 5]
```diff
@@ -189,6 +189,30 @@ <h5>
...
                 </div>
             </div>
+            <div class="col mb-4">
+                <div class="card">
+                    <div class="card-header">
+                        <h5>
+                            <i class="fas fa-calendar"></i>
+                            Employee Scheduling
+                            <!-- TODO: Blog Post -->
+                            <a href="https://www.optaplanner.org/blog/2021/10/14/OrderPickingQuickstart.html" target="_blank" title="About this quickstart" class="float-right">
```
Comment (ge0ffrey) [inline build/quickstarts-showcase/src/main/resources/META-INF/resources/index.html]: Let's use https://www.optaplanner.org/learn/useCases/employeeRostering.html - the employee scheduling landing page for now

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/bootstrap/DemoDataGenerator.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    EmployeeRepository employeeRepository;
+    @Inject
+    AvailabilityRepository availabilityRepository;
+    @Inject
+    ShiftRepository shiftRepository;
+    @Inject
+    ScheduleStateRepository scheduleStateRepository;
+
+    @Transactional
+    public void generateDemoData(@Observes StartupEvent startupEvent) {
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/bootstrap/DemoDataGenerator.java:94]: Can we make the shift generation more realistic: 90% or 80% predictable, and 10% to 20% random?

Comment (ge0ffrey) [reply use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/bootstrap/DemoDataGenerator.java:94]: image

Comment (ge0ffrey) [reply use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/bootstrap/DemoDataGenerator.java:94]: General idea:
For each location, roll a dice if it's has, for every day by default:
- 2 shifts (early + late)
- 3 shifts (early, late, night)
- 4 shifts (early, late, night, day)

Then per location, add in a few extra exceptional shifts on some dates (but not all). For example: add 1 or 2 extra day shifts for a few dates (for example per day 80% none, 10% 1 extra shift, 10% 2 extra day shifts). Or simulate a festival/sport event in town day which adds in an extra late and night shift on that day.

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Availability.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+@Entity
+public class Availability {
+
+    @Id
+    @PlanningId
+    @GeneratedValue
+    Long id;
+
+    @ManyToOne(optional = false)
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Availability.java]: We don't bother to do optional=false and other JPA details in our quickstarts, to make them easier to read.

Originally, we even added hibernate-validator to do @NotNull etc. But for quickstarts, it's just makes it a slower start for users. We don't need to teach them hibernate-jpa. So let's remove that distraction

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Availability.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import javax.persistence.ManyToOne;
+
+import org.optaplanner.core.api.domain.lookup.PlanningId;
+
+@Entity
+public class Availability {
+
+    @Id
+    @PlanningId
+    @GeneratedValue
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Availability.java:17]: Nitpick: order of annotations is jpa - optaplanner - jpa. See other examples, they do:

```

    @PlanningId
    @Id
    @GeneratedValue
    private Long id;
```

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Availability.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public String toString() {
+        return "Availability{" +
+                "id=" + id +
+                ", employee=" + employee +
+                ", date=" + date +
+                ", availabilityType=" + availabilityType +
+                '}';
+    }
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Availability.java:71]: The toString() of domain objects is used in optaplanner's DEBUG and TRACE logging, to explain the move.
This is too verbose, making those logs unreadable.

A good toString() is short and sweet. The least amount of information to identify the object uniquely (presuming an employee's name is unique). See other examples how to do that.

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Employee.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    public void setSkillSet(Set<String> skillSet) {
+        this.skillSet = skillSet;
+    }
+
+    @Override
+    public String toString() {
+        return "Employee{" +
+                "name='" + name + '\'' +
+                ", skillSet=" + skillSet +
+                '}';
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Employee.java]: In this case, it's easy. Just `return name;`.

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/EmployeeSchedule.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    List<Shift> shiftList;
+
+    @PlanningScore
+    HardSoftScore score;
+
+    ScheduleState scheduleState;
+
+    SolverStatus solverStatus;
+
+    public EmployeeSchedule() {}
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/EmployeeSchedule.java:33]: See other examples. We typically add a comment here to explain that OptaPlanner needs a no-arg construction for planning solution classes.

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/ScheduleState.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+@Entity
+public class ScheduleState {
+
+    @Id
+    Long tenantId;
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/ScheduleState.java:16]: No @GeneratedValue?

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/ScheduleState.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    @JsonIgnore
+    public LocalDate getFirstPublishedDate() {
+        return lastHistoricDate.plusDays(1);
+    }
+
+    public LocalDate getFirstUnplannedDate() {
+        return firstDraftDate.plusDays(draftLength);
+    }
+
+    // Simple getters/setters
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/ScheduleState.java]: Nitpick. For clarity and consistency, can we do it like the other examples?


```
    // ************************************************************************
    // Getters and setters
    // ************************************************************************
```

File: use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Shift.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    @Override
+    public String toString() {
+        return "Shift{" +
+                "id=" + id +
+                ", start=" + start +
+                ", end=" + end +
+                ", location='" + location + '\'' +
+                ", requiredSkill='" + requiredSkill + '\'' +
+                ", employee=" + employee +
+                '}';
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/java/org/acme/employeescheduling/domain/Shift.java]: Here, the employee would definitly not be part of it (the DEBUG/TRACE message toStrings the old and new employee of the move evaluated). What's the minimum we need? Maybe `return location + " " + start + "-" + end;`.

File: use-cases/employee-scheduling/src/main/resources/META-INF/resources/app.js [hunk 15]
```diff
@@ -0,0 +1,361 @@
+let autoRefreshIntervalId = null;
+const zoomMin = 2 * 1000 * 60 * 60 * 24 // 2 day in milliseconds
+const zoomMax = 4 * 7 * 1000 * 60 * 60 * 24 // 4 weeks in milliseconds
+
+const byEmployeePanel = document.getElementById("byEmployeePanel");
+const byEmployeeTimelineOptions = {
+    timeAxis: {scale: "day"},
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/resources/META-INF/resources/app.js]: Let's show hours too, as that matters for this use case when a shift starts

File: use-cases/employee-scheduling/src/main/resources/META-INF/resources/index.html [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        <div class="float-right">
+            <ul class="nav nav-pills" id="myTab" role="tablist">
+                <li class="nav-item active">
+                    <a class="nav-link" id="byLocationPanelTab" data-toggle="tab" href="#byLocationPanel" role="tab"
+                       aria-controls="byLocationPanel" aria-selected="true">By location</a>
+                </li>
+                <li class="nav-item">
+                    <a class="nav-link" id="byEmployeePanelTab" data-toggle="tab" href="#byEmployeePanel" role="tab"
+                        aria-controls="byEmployeePanel" aria-selected="false">By employee</a>
+                </li>
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/resources/META-INF/resources/index.html:74]: One of these should be active by default. If I start up, the tab that is active doesn't look that way. It's a matter of adding a CSS class for the active one, see other examples.

File: use-cases/employee-scheduling/src/main/resources/META-INF/resources/index.html [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+        <button id="stopSolvingButton" type="button" class="btn btn-danger">
+            <span class="fas fa-stop"></span> Stop solving
+        </button>
+        <button id="publish" type="button" class="btn btn-secondary">
+            <span class="fas fa-refresh"></span> Publish
+        </button>
+        <span id="unassignedShifts" class="ml-2 align-middle font-weight-bold"></span>
+        <span id="score" class="score ml-2 align-middle font-weight-bold">Score: ?</span>
+
+        <div class="float-right">
```
Comment (ge0ffrey) [inline use-cases/employee-scheduling/src/main/resources/META-INF/resources/index.html:65]: Tab location is off when I run it locally:

image

Comment (ge0ffrey) [reply use-cases/employee-scheduling/src/main/resources/META-INF/resources/index.html:65]: Tabs should appear above the schedule, like in maintenance scheduling example.

**Review by ge0ffrey — APPROVED** [review]
Great work. Take some time to do the data generator more realistic, but other than that, good to merge!

**Review by ge0ffrey — APPROVED** [review]

**Comment by ge0ffrey** [comment]
Great work Chris!

**Pull request merged**